site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+}
+
+func (c TraceAnimal) Speak(x int) {
+	fmt.Printf("Running Speak(x) function with x=%d!\n",x)
+	c.IAnimal.Speak(x)
+}
+
+func (c MeasureAnimal) Speak(x int) {
+	fmt.Println("Timing Speak() function...")
+	t := time.Now()
+	c.IAnimal.Speak(x)
+	
+	fmt.Printf("Speak(%d) took %s\n", x, time.Since(t))
+}
+
+func main() {
+	a := Animal{Name: "Cow"}
+	c := TraceAnimal {IAnimal: a}
+	d := MeasureAnimal{IAnimal: c}
+	d.Speak(2)
+}
+```
+
+Running the following code will yield:
+
+```
+Timing Speak() function...
+Running Speak(x) function with x=2!
+I am a Cow I am a Cow 
+Speak(2) took 0s
+```
+
+So we've basically implemented two decorators over original `Speak()` method, first we started timing the execution in `MeasureAnimal`, then passed it too `TraceAnimal` which in turn called the actual `Speak()` implementation.
+
+This works great and stays performant since we don't use any dynamic techniques here like `reflection`, however this is very verbose and requires us to write a lot of wrapper code, and that's no fun at all. We can do better!
+
+### Code parsing using AST
+
+Using the things we've discussed in parts [1]({{< relref "./2019-10-25-instrumenting-go-code-via-ast.md" >}}) & [2]({{< relref "./2020-03-15-instrumenting-go-code-via-ast-2.md" >}}) of these series we can scan the interface of the struct we want to wrap and collect all the information needed to generate the decorators/wrappers automatically. Let's dig in.
+
+First of all, we kick of AST parser on our input file that contains the interface and start walking through the found nodes:
+
+```go
+	fset := token.NewFileSet() // positions are relative to fset
+
+	file, err := os.Open("animal.go")
+	if err != nil {
+		return nil, fmt.Errorf("Unable to open %s file: %w", inputFile, err)
+	}
+	src, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	f, err := parser.ParseFile(fset, "animal.go", src, parser.AllErrors|parser.ParseComments)
+	if err != nil {
+		return nil, err
+	}
+
+	ast.Inspect(f, func(n ast.Node) bool {
+		// ... handle the found nodes
+	})
+```
+
+To differentiate interface methods from other AST nodes, we can do the following: 
+
+```go
+	ast.Inspect(f, func(n ast.Node) bool {
+		switch x := n.(type) {
+		case *ast.TypeSpec:
+			if x.Name.Name == "IAnimal" {
+				for _, method := range x.Type.(*ast.InterfaceType).Methods.List {
+					methodName := method.Names[0].Name
+					// here we can parse all the required information about the method
+					methods[methodName] = extractMethodMetadata(method, src)
+				}
+			}
+		}
+		return true
+	})
+```
+
+Let's define a couple of struct to help us collect information about methods:
+
+```go
+type methodParam struct {
+	Name string
+	Type string
+}
+
+type methodData struct {
+	Params        []methodParam
+	Results       []string
+}
+
+
+// for each found method, we'll store it's name, params with their types and return types in methods := map[string]methodData {}
```

Comment:
```suggestion
// For each found method we'll store its name, params with their types, and return types in methods := map[string]methodData {}
```

Reply:
@reflog, I'm not sure I punctuated this comment properly in terms of making name, types, types in methods being an actual list. Are those three items part of the found method?

Reply:
Yep. Your corrections are spot on